docs/reference/development-guide.rst

```diff
@@ -0,0 +1,412 @@
+the following actions take place for each variant:
+
+* ``omnitrace-causal`` calls ``fork()``
+* the child process launches ``<command> <command-args>`` using ``execvpe``, which modifies the environment for the variant
+* the parent process waits for the child process to finish
+
+omnitrace-instrument: `source/bin/omnitrace-instrument <https://github.com/ROCm/omnitrace/tree/main/source/bin/omnitrace-instrument>`_
+-------------------------------------------------------------------------------------------------------------------------------------------
+
+* Requires a command-line format of ``omnitrace-instrument <options> -- <command> <command-args>``
+* Allows the user to provide options specifying whether to perform runtime instrumentation, use binary rewrite, or 
+  attach to process
+* Either opens the instrumentation target (for binary rewrite), launches the target and stops it
+  before it starts executing ``main``, or attaches to a running executable and pauses it
+* Finds all functions in the targets
+* Finds ``libomnitrace-dl`` and locates the functions
+* Iterates over and instruments all the functions, provided they satisfy the 
+  defined criteria (such as a minimum number of instructions)
+
+  * See the ``module_function`` class
+
+* Until this point, the workflow has been the same for the different options, 
+  but it diverges after instrumentation is complete:
+
+  * For a binary rewrite: it produces a new instrumented binary and exits
+  * For runtime instrumentation or attaching to a process: it instructs the application 
+    to resume and then waits for it to exit
+
+Libraries
+========================================
+
+Common library: `source/lib/common <https://github.com/ROCm/omnitrace/tree/main/source/lib/common>`_
+--------------------------------------------------------------------------------------------------------------------------------
+
+* General header-only functionality used in multiple executables and/or libraries. 
+* Not installed or exported outside of the build tree.
+
+Core library: `source/lib/core <https://github.com/ROCm/omnitrace/tree/main/source/lib/core>`_
+--------------------------------------------------------------------------------------------------------------------------------
+
+* Static PIC library with functionality that does not depend on any components. 
+* Not installed or exported outside of the build tree.
+
+Binary library: `source/lib/binary <https://github.com/ROCm/omnitrace/tree/main/source/lib/binary>`_
+--------------------------------------------------------------------------------------------------------------------------------
+
+* Static PIC library with functionality for reading/analyzing binary info.
+* Mostly used by the causal profiling sections of ``libomnitrace``.
+* Not installed or exported outside of the build tree.
+
+libomnitrace: `source/lib/omnitrace <https://github.com/ROCm/omnitrace/tree/main/source/lib/omnitrace>`_
```

Comment:
Should these links be updated to point to the release branch? Or changed to use relative links?

Reply:
Right now, they seem to go to the "main" branch in GitHub. Users can select a branch from there. But I'll ask around about this to find out how this is handled elsewhere.

Reply:
Currently only links to other documentation via intersphinx (https://rocm.docs.amd.com/projects/rocm-docs-core/en/latest/user_guide/linking.html) are able to take the branch as a parameter (ie: 6.0 links to other 6.0)

Doing the same for links to GitHub is still an in-progress feature

Reply:
In that case, should we change to <https://github.com/ROCm/omnitrace/tree/release/rocm-rel-6.2/source/lib/omnitrace>.
And similarly change other links from `/tree/main` to `tree/release/rocm-rel-6.2`.

Reply:
The one downside of that is that we have to remember to update the links with every release. So if we don't update the links for 6.3, then they will point to 6.2. But I'm open to discussion on this.